nucleus/common/glassfish-api/src/main/java/org/glassfish/api/jdbc/SQLTraceRecord.java

```diff
@@ -194,21 +214,111 @@ public void setParams(Object[] params) {
         this.params = params;
     }
 
+    /**
+     * SQL query related to the database operation if applicable.
+     *
+     * @return SQL query
+     */
+    public Optional<String> getSqlQuery() {
+        return Optional.ofNullable(sqlQuery);
+    }
+
+    /**
+     * Set the SQL query related to the database operation
+     *
+     * @param sqlQuery
+     */
+    public void setSqlQuery(String sqlQuery) {
+        this.sqlQuery = sqlQuery;
+    }
+
+    /**
+     * Application Name which executed the SQL statement
+     *
+     * @return Application name
+     */
+    public String getApplicationName() {
+        return applicationName;
+    }
+
+    /**
+     * Set the application name which executed the SQL statement
+     *
+     * @param applicationName
+     */
+    public void setApplicationName(String applicationName) {
+        this.applicationName = applicationName;
+    }
+
+    /**
+     * Module name of an application which executed the SQL statement.
+     * If the application doesn't have modules it will be equal to the application name
+     *
+     * @return Module name
+     */
+    public String getModuleName() {
+        return moduleName;
+    }
+
+    /**
+     * Set the module name which executed the SQL statement
+     *
+     * @param moduleName
+     */
+    public void setModuleName(String moduleName) {
+        this.moduleName = moduleName;
+    }
+
+    /**
+     * Get the stack trace element for the method call in the application that triggered SQL execution.
+     *
+     * This call analyzes the stacktrace of the current thread and finds the first element that
+     * represents a call in the application (the class is loaded by the application classloader).
+     *
+     * @return Stack trace element that represents a call to a server component that triggered SQL execution
+     */
+    public Optional<StackTraceElement> getCallingApplicationMethod() {
+        return Arrays.stream(Thread.currentThread().getStackTrace())
```

Comment:
Why not use `StackWalker` API with the `RETAIN_CLASS_REFERENCE`?

Reply:
Thank you for the suggestion, I forgot about this API. I redesigned my solution using StackWalker and moved the computation outside of the SQLTraceListener. It means that it will be computed even if the listener doesn't need it. I think it's OK because now the computation is more efficient. It's still computed only if there is at least 1 listener.